pkg/providers/amifamily/bootstrap/nodeadm.go

```diff
@@ -0,0 +1,152 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	admapi "github.com/awslabs/amazon-eks-ami/nodeadm/api"
+	admv1alpha1 "github.com/awslabs/amazon-eks-ami/nodeadm/api/v1alpha1"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily/bootstrap/mime"
+)
+
+type Nodeadm struct {
+	Options
+}
+
+func (n Nodeadm) Script() (string, error) {
+	nodeConfigYAML, err := n.getNodeConfigYAML()
+	if err != nil {
+		return "", fmt.Errorf("generating NodeConfig, %w", err)
+	}
+	customEntries, err := n.parseUserData()
+	if err != nil {
+		return "", fmt.Errorf("parsing custom UserData, %w", err)
+	}
+	mimeArchive := mime.Archive(append([]mime.Entry{{
+		ContentType: mime.ContentTypeNodeConfig,
+		Content:     nodeConfigYAML,
+	}}, customEntries...))
+	userData, err := mimeArchive.Serialize()
+	if err != nil {
+		return "", err
+	}
+	return userData, nil
+}
+
+// getNodeConfigYAML returns the Karpenter generated NodeConfig YAML object serialized as a string
+func (n Nodeadm) getNodeConfigYAML() (string, error) {
+	config := &admv1alpha1.NodeConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       admapi.KindNodeConfig,
+			APIVersion: admv1alpha1.GroupVersion.String(),
+		},
+		Spec: admv1alpha1.NodeConfigSpec{
+			Cluster: admv1alpha1.ClusterDetails{
+				Name:              n.ClusterName,
+				APIServerEndpoint: n.ClusterEndpoint,
+			},
+		},
+	}
+	if cidr := lo.FromPtr(n.ClusterCIDR); cidr != "" {
+		config.Spec.Cluster.CIDR = cidr
+	} else {
+		return "", fmt.Errorf("resolving cluster CIDR")
```

Comment:
If this fails to be resolved, I think that we should be emitting a NodeClassNotReady error rather than a raw error. The reason is that we know that this NodeClaim relies on the EC2NodeClass resolution of this value, it's possible that the value isn't resolved yet by the time that it's accessed here, so  you need to handle that eventual consistency some way